tools/meteor-services/deploy.js

```diff
@@ -799,6 +814,88 @@ var listSites = function () {
   return 0;
 };
 
+// Given a hostname, add "http://" or "https://" as
+// appropriate. (localhost gets http; anything else is always https.)
+function addScheme(hostOrURL) {
+  if (hostOrURL.match(/^http/)) {
+    return hostOrURL;
+  } else if (hostOrURL.match(/^localhost(:\d+)?$/)) {
+    return "http://" + hostOrURL;
+  } else {
+    return "https://" + hostOrURL;
+  }
+};
+
+// Maps from "site" to deploy URL, so we don't have to re-ping on each RPC.
+const galaxyDiscoveryCache = new Map;
+
+// getDeployURL returns the base deploy URL for the given app.  "app" may be
+// falsey for certain RPCs (eg meteor list-sites).
+function getDeployURL(site) {
+  // Always trust explicitly configuration via env.
+  if (process.env.DEPLOY_HOSTNAME) {
+    return addScheme(process.env.DEPLOY_HOSTNAME);
+  }
+
+  const defaultURL = "https://galaxy.meteor.com";
+
+  // No site? Just use the default.
+  if (!site) {
+    return defaultURL;
+  }
+
+  // If we have a site, we can try to do Galaxy discovery.
+
+  // Do we already have an answer?
+  if (galaxyDiscoveryCache.has(site)) {
+    return galaxyDiscoveryCache.get(site);
+  }
+
+  // Try https first.  Production users should enable https so we will never
+  // fall back to http for them.
+  const urlFromHTTPS = discoverGalaxy(site, "https");
```

Comment:
Thanks, I appreciate the idea to use Promises (and your help on Slack).  This really does end up a lot cleaner!  I've force pushed (only editing the final commit) tweaking this.  Please re-review, including any advice on my rusty JS.